spec/data-model/README.md

```diff
@@ -25,33 +25,79 @@ Two equivalent definitions of the data model are also provided:
 A `SelectMessage` corresponds to a syntax message that includes _selectors_.
 A message without _selectors_ and with a single _pattern_ is represented by a `PatternMessage`.
 
+In the syntax,
+a `PatternMessage` may be represented either as a _simple message_ or as a _complex message_,
+depending on whether it has declarations and if its `pattern` is allowed in a _simple message_.
+
+While this specification currently only supports `"match"` selectors,
+the _matcher_ syntax may include a _reserved statement_,
+allowing for future expansion.
+In the data model, this is represented as a `SelectMessage` with a string `type`
+that is not `"input"`, `"local"`, `"match"`, or `"message"`
+(corresponding to a concatenation of the `UnsupportedStatement` `keyword` and `body`),
+and with `selectors` matching the _reserved statement_ _expressions_.
```

Comment:
Hang on, let's rewind a bit. I think there's an issue in the current ABNF this is pointing at. Consider this message:
```
.foo {$var}
.foo {$var}
* {{text}}
```

I believe the _intent_ of our current ABNF would have the first `.foo {$var}` parse as a `declaration`, but to have the second one end up in the `matcher`. But that can only happen if we allow for infinite lookahead past the `1*([s] expression)` and for the ambiguity to be resolved by only one overall interpretation of the message providing a valid parse.

And I don't think we'd be happy with that.

I can imagine two different approaches to resolve this:

1. Rather than having something like `declarations`, have `statements` (including `.match`) at the start of a complex message. Then make having `match-statement` or `reserved-statement` be the last one (if present) be a requirement for it to be _valid_ rather than _well-formed_.
2. Make the selector syntax distinct from what's allowed in `declarations`, e.g. with `.match [{$foo}]`.

Option 1 also makes it much harder to allow for statements like `.strict true`, because we need to detect the end of the selectors.

Option 2 adds syntactical noise, and would still require lookup past the `reserved-statement` keyword to disambiguate the syntax.

And what do we get out of it? An ability to later add something like `.select` as an alternative to `.match`. Given that we've already delegated _a lot_ to custom selectors, and we're not attempting to remove the validity requirements of selector/variant-keys count matching or having an all-catch-all case, how could this `.select` differ from `.match`? The only remaining aspect that I can imagine is using first-match rather than best-match for the selection.

@aphillips @stasm Would you concur that the cost/benefit ratio here is not very good, and that we should drop `reserved-statement` from the `matcher`? I think first-match vs best-match is something that we could assess post-tech-preview and, if necessary, apply as a behaviour change on `.match` rather than via a later keyword introduction.

Reply:
My thinking here was that building too much flexibility via the reserved syntax was powerful but also dangerous -- we may end up with a very different syntax than right now. Designating some keywords as invariant, like `.match`, seems like a good idea.

The first-match `.select` use-case gives me pause. Maybe it would be best to explicitly reserve `.select` as the 4th keyword?

The entire problem of forward compatibility and reserved syntax is complex enough that perhaps it would make sense to document it in a design doc? What sort of extensions do we want to make possible? Is there anything that we can rule out today? How do build all of that into the ABNF?

@gibson042 Would you be interested in writing such a design?

Reply:
@aphillips What do you think of the idea of including `.select` as a reserved alternative to `.match`, otherwise using the same syntax? In the ABNF it would probably look like this:
```abnf
matcher = match-statement 1*([s] variant)
match-statement = (match / select) 1*([s] selector)
match = %s".match"
select = %s".select"
```

As with other reserved syntax, this would generate an error during formatting, and not in the data model.

The intent here would be to allow for one future opportunity to iterate on `.match`, should we identify any issues with it after the 2.0 release. In terms of process steps, its [selection method](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#pattern-selection) is by far the most complex part of the spec.

Reply:
I think I'd be against it, particularly if we also had a `:select` matcher. 

I know I brought this up originally, but I think having multiple selection mechanisms isn't helpful for users (if `.select` were a real keyword there would be a constant question of which matcher to use whether the results were "correct" for a given one). 

I agree that the best-match selection is the most complex part of the spec, but it is consistent (or should be) with what MF1 already does. We have a lot of experience with that. If we need to adjust the algorithm because of something we find in testing, I'm at peace with that. But I think it will not be helpful to have more kinds of selection.

Clarification of my earlier comment:
> The extension mechanism for .match could use one of the forest of reserved sigils rather than different keywords.

What I meant here was replacing functions, not a different sigil for the keyword, e.g. `.match {$var ^extended}`

So, in summary, I favor:

1. Remove reserved-statement from the selectors
2. Stick with the current matcher
3. Be done

Reply:
Let's drop it then.